microsoft/testsuites/dpdk/dpdktestpmd.py

```diff
@@ -264,32 +264,49 @@ def generate_testpmd_command(
             "-a --stats-period 1"
         )
 
+    def generate_testpmd_killer(self, signal: str = "KILL") -> str:
+        return f"killall -s {signal} {self.command}"
+
     def run_for_n_seconds(self, cmd: str, timeout: int) -> str:
         self._last_run_timeout = timeout
         self.node.log.info(f"{self.node.name} running: {cmd}")
-        self.timer_proc = self.node.execute_async(
-            f"sleep {timeout} && killall -s INT {cmd.split()[0]}",
+
+        # testpmd exits cleanly when killed with INT.
+        # setup some timers to kill it after timeout
+        self.node.execute_async(
+            f"sleep {timeout} && {self.generate_testpmd_killer()}",
```

Comment:
Can you please check if we can do something like this:
1. Use timeout on the test case
2. Add the kill in the `after_case` method?

Reply:
Worth checking, I think that I had issues with the shell.kill command that timeout will call but maybe that assumption has changed.

Reply:
If sleep before kill, it may make trouble. The kill is called after certain seconds, it may kill the wanted processes at that time.

Reply:
Ah I remember, using wait_result(timeout=20) results in:
`2022-06-09 00:50:08.481[2972][DEBUG] lisa.env[generated_0].node[0].cmd[2820] failed on killing process: return code: 1
output: b''
stderr output: b'kill: (10366): Operation not permitted\n'`

and the test ends up failing because stdout from the process isn't available in the result object after the failure.

Reply:
node.shell.kill seems broken :\ maybe we can point it to the kill tool instead?

Reply:
We'll fix that in another PR

Reply:
@somil55 Ah I forgot to clarify, the scheme here isn't to clean up. It's so we can stop the program and collect the test results after a variable amount of time. There is still a max timeout on the test case itself that will kill everything if things stall for some reason.